### ISSUE TITLE
Support restore priority for some namespace resources over some cluster resources.

### ISSUE THREAD

**Issue by dgoodwin** [heptio/ark#111]
Original problem surfaces as a restore error on OpenShift with:

```
error restoring /tmp/553128096/cluster/oauthaccesstokens.oauth.openshift.io/K8QS3VweprsrYqACUwi5mz3DOaR0PWiCetX4kZVIuLE.json:                                                                            
     OAuthAccessToken "K8QS3VweprsrYqACUwi5mz3DOaR0PWiCetX4kZVIuLE" is invalid: clientName:                      
      Internal error: serviceaccounts "jenkins" not found'
```

A second restore will correct the problem as the first one created the dependent service account.

In this case we need service accounts restored before these cluster scoped oauthaccesstokens. However, adding these to the resource priorities does not work as all cluster resources are processed first. See: https://github.com/heptio/ark/blob/v0.4.0/pkg/restore/restore.go#L272

May need some kind of system for processing namespace scoped resources before some cluster scoped.

**Comment by ncdc** [comment]
@skriss I'm thinking it may be easiest to change the backup tarball format to store first by resource, then by namespace. This will allow us to appropriately prioritize all resources, regardless of cluster- vs namespace-scoped.

Otherwise, we would need to do 2 passes through the `cluster` and `namespaces` directories:
1. Pass 1: handle prioritized resources
1. Pass 2: handle everything else

Sorting first by resource does have a downside: it makes it less efficient to restore a subset of namespaces. We currently don't descend into an individual namespace directory (and all its resource directories) if it's excluded. With this change, we would read every resource directory. If a given namespace only had a couple of resources in it, we'd end up reading a lot more directories than the current implementation. I'm not sure that's a big enough detractor, though.

**Comment by ncdc** [comment]
If we do make the change to the backup format, I also want to consider including the `Backup` in the tarball (so we have the snapshot info). And I definitely want to make sure the backup version is included in a file in the tarball (either in the Backup resource, or separately).

**Comment by dgoodwin** [comment]
The oauth access token in question:

```
{
    "apiVersion": "oauth.openshift.io/v1",
    "authorizeToken": "",
    "clientName": "system:serviceaccount:jenkins:jenkins",
    "expiresIn": 86400,
    "kind": "OAuthAccessToken",
    "metadata": {
        "creationTimestamp": "2017-09-29T17:42:31Z",
        "name": "K8QS3VweprsrYqACUwi5mz3DOaR0PWiCetX4kZVIuLE",
        "resourceVersion": "31791",
        "selfLink": "/apis/oauth.openshift.io/v1/oauthaccesstokens/K8QS3VweprsrYqACUwi5mz3DOaR0PWiCetX4kZVIuLE",
        "uid": "91a066d8-a53d-11e7-a3b0-0e0185b64950"
    },
    "redirectURI": "https://jenkins-jenkins.54.157.56.222.example.com/securityRealm/finishLogin",
    "scopes": [
        "user:info",
        "user:check-access"
    ],
    "userName": "admin",
    "userUID": "9ba4e317-a50d-11e7-a3b0-0e0185b64950"
}
```

**Comment by ncdc** [comment]
@dgoodwin and the jenkins SA has no reference to this resource?

**Comment by dgoodwin** [comment]
Service account looks like this:

```
{                                                        
    "apiVersion": "v1",                                                                                            
    "imagePullSecrets": [                                                                                          
        {                                                
            "name": "jenkins-dockercfg-nfqx2"            
        }                                                                                                          
    ],                                                   
    "kind": "ServiceAccount",                                                                                      
    "metadata": {
        "annotations": {
            "serviceaccounts.openshift.io/oauth-redirectreference.jenkins": "{\"kind\":\"OAuthRedirectReference\",\"apiVersion\":\"v1\",\"reference\":{\"kind\":\"Route\",\"name\":\"jenkins\"}}"
        },
        "creationTimestamp": "2017-09-29T17:36:41Z",
        "labels": {
            "app": "jenkins-ephemeral",
            "template": "jenkins-ephemeral-template"
        },
        "name": "jenkins",
        "namespace": "jenkins",
        "resourceVersion": "31264",
        "selfLink": "/api/v1/namespaces/jenkins/serviceaccounts/jenkins",
        "uid": "c0edd52b-a53c-11e7-a3b0-0e0185b64950"
    },
    "secrets": [
        {
            "name": "jenkins-dockercfg-nfqx2"
        },
        {
            "name": "jenkins-token-6hk6g"
        }
    ]
}
```

**Comment by dgoodwin** [comment]
Regarding storing first by resource then by namespace, am I understanding that this would mean:

```
cluster/securitycontextconstraints/hostaccess.json
cluster/namespaces/logging.json
namespaces/jenkins/serviceaccounts/builder.json
```

Would become:

```
securitycontextconstraints/hostaccess.json
namespaces/logging.json
serviceaccounts/jenkins/builder.json
```

Possibly with all of those in a "resources/" sub-dir or similar to isolate them from any other data in the backup?

Then the current mechanism for prioritizing and ordering resources would apply across both cluster level and namespace level resources.

**Comment by ncdc** [comment]
I think something like this:
```
/resources/
  namespaces/
    foo.json
  pods/
    foo/
      my-pod.json
  storageclasses.storage.k8s.io/
    my-storage-class.json
```

That might make it slightly challenging to know if we're dealing with a cluster- or namespace-scoped resource when we're walking the directory tree. We will likely either need some slight modifications to that structure to make that easier, or we could write out a file such as `.ark` in each resource directory that contains relevant metadata (cluster or namespace scoped, and anything else we may need).

**Comment by ncdc** [comment]
And yes +1 to prioritizing across all resources, regardless of scope

**Comment by dgoodwin** [comment]
Would pods/namespaces/foo/ and storageclasses.storage.k8s.io/cluster/ make things more clear?

**Comment by ncdc** [comment]
Maybe pods/foo/pod.json and storageclasses.storage.k8s.io/cluster/sc.json. That way the nesting level is always the same?

**Comment by dgoodwin** [comment]
Sounds good to me. Working on this now.

**Comment by dgoodwin** [comment]
Might be a problem if someone used a namespace "cluster", do we want/need to be able to look at a dir and know if it's cluster or namespace scoped just from the file structure?

**Comment by ncdc** [comment]
That is one of my concerns, and the reason I suggested a `.ark` metadata file.

If we went with your suggestion, for a given resource directory, we could check for the existence of either "namespaces" or "cluster", and whatever existed would indicate which it is. Maybe that's better?

**Comment by dgoodwin** [comment]
Not crazy about the differing levels but I think I'd trade it for being able to eyeball it and know implicitly what I'm looking at, I'd go with pods/namespaces/foo/pod.json I think.

**Comment by ncdc** [comment]
@skriss any opinions?

**Comment by ncdc** [comment]
Fixed by #132
